src/app/admin/product-form/product-form.component.ts

```diff
@@ -1,25 +1,49 @@
 import { Component } from '@angular/core';
 import { CategoryService } from 'src/app/category.service';
-import { Observable } from 'rxjs';
+import { Observable, pipe } from 'rxjs';
 import { ProductService } from 'src/app/product.service';
-import { Router } from '@angular/router';
+import { Router, ActivatedRoute } from '@angular/router';
+import { take } from 'rxjs/operators';
 
 @Component({
   selector: 'app-product-form',
   templateUrl: './product-form.component.html'
 })
 export class ProductFormComponent {
   categoriesObservable: Observable<any[]>;
+  product = {};
+  id;
   constructor(
     private router: Router,
-    private categoryService: CategoryService,
+    private route: ActivatedRoute,
+    categoryService: CategoryService,
     private productService: ProductService
   ) {
     this.categoriesObservable = categoryService.getCategories();
+    this.id = this.route.snapshot.paramMap.get('id');
+    if (this.id)
+      this.productService
+        .get(this.id)
+        .pipe(take(1))
```

Comment:
can you put this pipe inside of the service and use a pipe in the html? that would get rid of the need to unsubscribe which we arent doing here